rfcs/0004-eslint-config-import.md

```diff
@@ -0,0 +1,307 @@
+- Start Date: 2020-01-20
+- Project: `@moxy/eslint-config-*`
+- Reference Issues: [#79](https://github.com/moxystudio/eslint-config/pull/79)
+- Implementation PR: (leave this empty)
+
+# Summary
+
+Add import rules to our code style so that we can automate some import order rules and standardize code practices, hence reducing code review issues.
+Effectively, it would make the order and grouping of the imports deterministic.
+
+# Motivation
+
+Across projects, different approaches for ordering and organizing import statements in js files are used. This creates code inconsistency and non-standard code styling.
+
+**Examples**
+
+
+[Header.js](https://gitlab.com/moxystudio/thu/thu-japan-2020/blob/master/www/shared/components/header/Header.js)
+
+```js
+import React from 'react';
+import PropTypes from 'prop-types';
+import classNames from 'classnames';
+// Components
+import { LanguageSelector } from './components';
+import SoundToggle from '../sound-toggle';
+// Icons
+import { LogoIcon } from '../icons';
+// Styles
+import styles from './Header.css';
+```
+
+[App.js](https://gitlab.com/moxystudio/thu/thu-japan-2020/blob/master/www/app/App.js)
+
+```js
+import React from 'react';
+import NextApp from 'next/app';
+import Head from 'next/head';
+import keyboardOnlyOutlines from 'keyboard-only-outlines';
+import registerGoogleTracking from './ga-tracking';
+import { withNextIntlSetup } from '@moxy/next-intl';
+import nextIntlConfig from '../../intl';
+
+import favicon16 from '../shared/media/favicons/favicon-16x16.png';
+import favicon32 from '../shared/media/favicons/favicon-32x32.png';
+import favicon96 from '../shared/media/favicons/favicon-96x96.png';
+import favicon192 from '../shared/media/favicons/favicon-192x192.png';
+
+import SEO_DATA from './App.data.js';
+
+import '../shared/styles/index.css';
+```
+
+Just like any other styling rule, ordering the import statements would benefit code-bases by making them consistent and reducing strain on code reviews and code style decisions.
+
+
+# Detailed design
+
+This RFC focuses on two new major styling proposals:
+
+ 1. Group order in import statements (dependencies first, relative imports after, etc) and order them accordingly.
+ 1. Within each import group, sort imports alphabetically
```

Comment:
Is this alphabetical rule mandatory?

Reply:
My reasoning is that I find it strange to have `classnames` before `react` in a React Component file 😅

Reply:
We could still have any React related import before anything else! We just need to use https://github.com/Tibfib/eslint-plugin-import-helpers.

Example:

```js
{
    // .eslintrc.js
    plugins: ['eslint-plugin-import-helpers'],
    rules: {
        'import-helpers/order-imports': [
            'warn',
            { // example configuration
                newlinesBetween: 'always',
                groups: [
					'/^react/',
                    'module',
                    ['parent', 'sibling', 'index'],
                ],
                alphabetize: { order: 'asc', ignoreCase: true },
            },
        ],
    }
}
```

Reply:
And just to be clear: I totally aggree with @paulobmarcos! It is really strange to have non-react related imports before react itself. I think we should only use this alphabetical rule if we also use the plugin I've just mentioned

Reply:
@guilhermedeandrade I was going to propose using that package as well but after looking at it more closely, I noticed that they [deprecated the `built-in` group](https://github.com/Tibfib/eslint-plugin-import-helpers/blob/master/docs/rules/order-imports.md#builtin--external--internal--module) and other groups in favor of one single group `module` forcing us to create a custom rule to be able to sort `built-in` modules in nodejs.
Because of this I don't like the direction the package is going.

Nevertheless, as mentioned in the rest of the RFC, an alternative would be to only sort by groups, and leave out the alphabetical sorting within groups.

Reply:
> Nevertheless, as mentioned in the rest of the RFC, an alternative would be to only sort by groups, and leave out the alphabetical sorting within groups.

Agree with this alternative! 👍

Reply:
> > Nevertheless, as mentioned in the rest of the RFC, an alternative would be to only sort by groups, and leave out the alphabetical sorting within groups.
> 
> Agree with this alternative! 👍

I also agree with this! 💯

Reply:
@zebateira If I remember correctly, there were some downsides of NOT sorting alphabetically. Or am I confusing with something else?

Reply:
I think it had to do with relative paths right?

Reply:
@satazor correct, not having alphabetic sorting means the imports order will not be deterministic and so some style variations will occur since you can sort within a group of imports.

Example of valid imports without alphabetical sorting:
```js
import fs from 'fs';
import path from 'path';

import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
```

```js
import path from 'path';
import fs from 'fs';

import React from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';
```


But it's better to at least have automatic group sort (without alpha sort) than none at all as I mentioned in the RFC.

Reply:
> I think it had to do with relative paths right?

Ah yes, there is also that "side effect". The rule [`import-plugin/order`](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/order.md#groups-array) within the group `parent` does not sort by folder depth, meaning `../a.js` and `../../../../b.js` fall within the same group. This means that if we enforce alphabetical order then we'd automatically have sorted imports by folder depth as a side effect as well.

Reply:
added

Reply:
will take this into consideration and remove the alphabetical order from the RFC 👍